### PR TITLE
[arch] Add service architecture docs

### PR DESCRIPTION
**Pull request by smithwjv**
* Adds documentation for how the services are architected.

This was quite tricky to get across clearly, hopefully I've achieved it. If there's anything you think should be in here but isn't let me know.

While writing this I felt like a big diagram could help, but when I came to designing one it felt convoluted and more likely to confuse than educate. If you can think of something helpful, again lmk.

### CODE REVIEW

File: TempleEight/docs/arch/service.md [hunk 1]
```diff
@@ -3,3 +3,155 @@ id: service
...
+## Service File Tree
+As a reminder, here is the `ExampleService` block from the [Getting Started](../getting-started.md) guide:
+```
+ExampleService: service {
+  foo: string;
+  bar: int;
+}
+```
+
+And here is the corresponding generated file tree:
```
Comment (jaylees14) [inline TempleEight/docs/arch/service.md:21]: Not a big fan of starting with And

File: TempleEight/docs/arch/service.md [hunk 2]
```diff
@@ -3,3 +3,155 @@ id: service
...
+
+We can ignore the `Dockerfile`, `go.mod` and `config.json` files for the purpose of illustrating the architecture.
+This leaves us with files that fit into one of three packages:
+
+ - `main` - sets up the server environment and handles incoming requests and outgoing responses
+ - `dao` - provides a common interface to the backing datastore, abstracting implementation details
+ - `util` - houses miscellaneous utility functions
+
+## The `main` Package
+This package, as you would expect, does most of the heavy lifting. In our example, it includes the following files:
```
Comment (jaylees14) [inline TempleEight/docs/arch/service.md:45]: Mention these files are located at the base directory level

File: TempleEight/docs/arch/service.md (hunk 1 above)
Comment (smithwjv) [reply TempleEight/docs/arch/service.md:21]: Neither, this was my attempt to write `casually`

File: TempleEight/docs/arch/service.md [hunk 3]
```diff
@@ -3,3 +3,173 @@ id: service
...
+
+```
+ExampleService: service {
+  foo: string;
+  bar: int;
+  another: AnotherExampleService;
+}
+
+AnotherExampleService: service {
+  baz: bool;
```
Comment (Lewiky) [inline TempleEight/docs/arch/service.md:120]: no fred 0/10

Comment (smithwjv) [reply TempleEight/docs/arch/service.md:120]: 😢

**Pull request merged**